src/merge_types.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import annotations
+
+from typing import List, Optional
+from pydantic import BaseModel, Field
+
+
+class SourceMetadata(BaseModel):
+    """Minimal source metadata available from extractions.
+
+    Fields are optional because many are not available from the current pipeline.
+    """
+
+    paper_id: str = Field(
+        ..., description="Identifier derived from the output filename (stem)"
+    )
+    title: Optional[str] = Field(default=None, description="Paper title if known")
+    section: Optional[str] = Field(default=None, description="Section heading if known")
+    paragraph_id: Optional[str] = Field(
+        default=None, description="Paragraph identifier if known"
+    )
+
+
+class LinkedEdgeSummary(BaseModel):
+    """Summarized view of an edge touching a node for LLM context."""
+
+    edge_type: str
+    rationale: str
+    confidence: float = Field(..., ge=0.0, le=1.0)
+    source: SourceMetadata
+
+
+class NodeAggregate(BaseModel):
+    """Aggregate information for a node across multiple papers/edges."""
+
+    node_key: str = Field(
+        ...,
+        description="Stable key used to identify this aggregate (e.g., canonical_name lowercased)",
+    )
+    text: str = Field(..., description="Primary surface name for the node (name)")
+    canonical_text: str = Field(
+        ..., description="Canonical name provided by extraction"
+    )
+    aliases: List[str] = Field(default_factory=list)
+    notes: List[str] = Field(
+        default_factory=list, description="Collected node notes across occurrences"
+    )
+    confidence_samples: List[float] = Field(default_factory=list)
+    linked_edges: List[LinkedEdgeSummary] = Field(default_factory=list)
+    sources: List[SourceMetadata] = Field(default_factory=list)
+
+
+class EdgeAggregate(BaseModel):
+    """Aggregate information for an edge across multiple papers."""
+
+    edge_key: str
+    edge_type: str
+    text: str = Field(..., description="Edge-level description if available")
+    source_nodes: List[str]
```

Comment:
Nit: Does it make sense to store them as tuples? Given (ConceptA, ConceptB) connected via EdgeA and (ConceptC, ConceptD) also connected through the same edge might allow ConceptA and ConceptD to be mapped even though relation doesn't exist.